src/postprocessing.py

```diff
@@ -200,4 +258,123 @@ def crop_image_center_per_class(image, h_crop, w_crop):
         cropped_prediction = class_prediction[h_start:-h_start, w_start:-w_start]
         cropped_per_class_prediction.append(cropped_prediction)
     cropped_per_class_prediction = np.stack(cropped_per_class_prediction)
-    return cropped_per_class_prediction
\ No newline at end of file
+    return cropped_per_class_prediction
+
+
+def join_score_image(image, score):
+    return (image, score)
+
+
+def get_features_for_image(image, probabilities, annotations):
+    image_features = []
+    category_layers_inds = np.cumsum(CATEGORY_LAYERS)
+    thresholds = get_thresholds()
+    for category_ind, category_instances in enumerate(image):
+        layer_features = []
+        category_nr = np.searchsorted(category_layers_inds, category_ind, side='right')
+        category_probabilities = probabilities[category_nr]
+        threshold = round(thresholds[category_ind], 2)
+        category_annotations = annotations.get(CATEGORY_IDS[category_nr], [])
+        iou_matrix = get_iou_matrix(category_instances, category_annotations)
+        for label_nr in range(1, category_instances.max() + 1):
+            mask = category_instances == label_nr
+            iou = get_iou(iou_matrix, label_nr)
+            mask_probabilities = np.where(mask, category_probabilities, 0)
+            area = np.count_nonzero(mask)
+            mean_prob = mask_probabilities.sum() / area
+            max_prob = mask_probabilities.max()
+            bbox = get_bbox(mask)
+            bbox_height = bbox[1] - bbox[0]
+            bbox_width = bbox[3] - bbox[2]
+            bbox_aspect_ratio = bbox_height / bbox_width
+            bbox_area = bbox_width * bbox_height
+            bbox_fill = area / bbox_area
+            dist_to_border = get_distance_to_border(bbox, mask.shape)
+            contour_length = get_contour_length(mask)
+            mask_features = {'iou': iou, 'threshold': threshold, 'area': area, 'mean_prob': mean_prob,
+                             'max_prob': max_prob, 'bbox_ar': bbox_aspect_ratio,
+                             'bbox_area': bbox_area, 'bbox_fill': bbox_fill, 'dist_to_border': dist_to_border,
+                             'contour_length': contour_length}
+            layer_features.append(mask_features)
+        image_features.append(pd.DataFrame(layer_features))
+    return image_features
+
+
+def get_iou_matrix(labels, annotations):
+    mask_anns = []
+    if annotations is None or annotations == []:
+        return None
+    else:
+        for annotation in annotations:
+            if not isinstance(annotation['segmentation'], dict):
+                annotation['segmentation'] = \
+                    cocomask.frPyObjects(annotation['segmentation'], labels.shape[0], labels.shape[1])[0]
+        annotations = [annotation['segmentation'] for annotation in annotations]
+        for label_nr in range(1, labels.max() + 1):
+            mask = labels == label_nr
+            mask_anns.append(rle_from_binary(mask.astype('uint8')))
```

Comment:
@apyskir I'd rather create `mask_ann =rle_...` and than mask_anns.append(mask_ann) .Easier to read

Reply:
@jakubczakon OK